LiteCore/Support/Logging.cc

```diff
@@ -294,24 +406,66 @@ namespace litecore {
         }
 
         // Write to the encoded log file:
-        if (sLogEncoder && level >= sFileMinLevel) {
-            sLogEncoder->vlog((int8_t)level, _name, (LogEncoder::ObjectRef)objRef, fmt, args);
+        if (level >= sFileMinLevel) {
+            dylog(level, _name, (LogEncoder::ObjectRef)objRef, fmt, args);
         }
     }
 
 
     void LogDomain::vlog(LogLevel level, const char *fmt, va_list args) {
-        vlog(level, LogEncoder::None, fmt, args);
+        vlog(level, LogEncoder::None, true, fmt, args);
     }
 
 
     void LogDomain::log(LogLevel level, const char *fmt, ...) {
         va_list args;
         va_start(args, fmt);
-        vlog(level, LogEncoder::None, fmt, args);
+        vlog(level, LogEncoder::None, true, fmt, args);
         va_end(args);
     }
 
+    void LogDomain::vlogNoCallback(LogLevel level, const char *fmt, va_list args) {
+        vlog(level, LogEncoder::None, false, fmt, args);
+    }
+
+
+    void LogDomain::logNoCallback(LogLevel level, const char *fmt, ...) {
+        va_list args;
+        va_start(args, fmt);
+        vlog(level, LogEncoder::None, false, fmt, args);
+        va_end(args);
+    }
+
+    void LogDomain::dylog(LogLevel level, const char* domain, unsigned objRef, const char *fmt, va_list args)
+    {
+        const char *objName = "?";
+        if (objRef) {
+            auto i = _objNames.find(objRef);
+            if (i != _objNames.end())
+                objName = i->second.c_str();
+        }
+
+        if(sLogEncoder[(int)level]) {
+            sLogEncoder[(int)level]->vlog(domain, (LogEncoder::ObjectRef)objRef, fmt, args);
+        } else if(sFileOut[(int)level]) {
+            static char formatBuffer[2048];
+            size_t n = 0;
+            LogDecoder::writeTimestamp(LogDecoder::now(), *sFileOut[(int)level]);
+            LogDecoder::writeHeader(kLevels[(int)level], domain, *sFileOut[(int)level]);
+            if (objRef)
+                n = snprintf(formatBuffer, sizeof(formatBuffer), "{%s#%u} ", objName, objRef);
+            vsnprintf(&formatBuffer[n], sizeof(formatBuffer) - n, fmt, args);
```

Comment:
Not sure if I'm pointing to the right spot or not. Based on my testing, log format from Objective-C that has %@ doesn't get formatted correctly. Can we use NSString to format the plain text log messages for Apple?

Reply:
Can you give an example of the logging line that does such a thing?  I thought that kind of stuff was built in already.  It seems like it would be following different logic since c4slog already takes a C4Slice which is supposed to be an already formatted UTF-8 string.

Reply:
iOS only calls c4vlog with the format that could include `%@` gets passed to the function directly. For callback log, the format will be passed back to the callback directly. Given https://github.com/couchbase/couchbase-lite-core/issues/631, is the right approach is to pass the already-formatted UTF-* string to c4vlog?

Reply:
Actually iOS can create a formatted UTF-8 string and use c4vlog(). We don't do a lot of logging from CBL side, probably that is OK.

Reply:
It might even be preferable.  Otherwise you are doing the formatting process three separate times, right?

Reply:
No, the formatting will be done twice (before calling c4log and in callback with "%s") instead of once in callback. I will ping you to see if there is a way to do it only once.

Reply:
As discussed with @borrrden, iOS will formatted the message before calling c4log so there will be no "%@" to worry about. This will inline with what the other platform is doing and the same thing as what @snej suggested in https://github.com/couchbase/couchbase-lite-ios/issues/2281.